### ISSUE TITLE
Unsupported keras layer type ZeroPadding2D

### ISSUE THREAD

**Issue by n-gregori** [deeplearning4j/deeplearning4j#2549]
hi, i'm trying to use the new functionality org.deeplearning4j.nn.modelimport.keras.Model , so i exported model_json and model_save with keras. in dl4j i found this exception: Exception in thread "main" org.deeplearning4j.nn.modelimport.keras.IncompatibleKerasConfigurationException: Unsupported keras layer type ZeroPadding2D. 
neuralNet used in this test is vgg16 : https://github.com/albertomontesg/keras-model-zoo/blob/master/kerasmodelzoo/models/vgg/vgg16.py
with the following steps:

in python:
model = vgg16.model(weights=True, summary=True)
model.compile(loss='mse', optimizer='sgd')
model.save_weights('vggw_save')
model_json = model.to_json()
with open("model.json", "w") as json_file:
    json_file.write(model_json)

in dl4j:
val model = Model.importSequentialModel("model.json","vggw_save")

with the exception:

17:11:22.940 [main] WARN  o.d.nn.modelimport.keras.Model - Changed layer-defined dropout 0.0 to 0.5 because of previous Dropout=0.5 layer
17:11:22.940 [main] WARN  o.d.nn.modelimport.keras.Model - Changed layer-defined dropout 0.0 to 0.5 because of previous Dropout=0.5 layer
Exception in thread "main" org.deeplearning4j.nn.modelimport.keras.IncompatibleKerasConfigurationException: Unsupported keras layer type ZeroPadding2D

**Comment by jadenyjw** [comment]
What's the progress on this issue?

**Comment by turambar** [comment]
@n-gregori @jadenyjw We have not yet added support for zero padding layers, but we fully support the [VGG model](https://github.com/fchollet/deep-learning-models/blob/master/vgg16.py) in the official Keras repository. If you're attempting to import a VGG-style model, we recommend using that  instead. If you have a non-VGG model that requires a separate zero padding layer, let us know -- and, if possible, share the model config with us.

**Comment by jadenyjw** [comment]
I'm using a non-VGG model, attempting to implement this [paper](https://arxiv.org/abs/1501.07873).
Here is the code that I've created in Keras defining my model:

`model = Sequential()

#L1
model.add(Convolution2D(nb_filter=64, nb_row=15, nb_col=15,
                        border_mode='valid',
                        input_shape=input_shape,
                        activation='relu',
                        subsample=(3,3)))
#model.add(ZeroPadding2D(padding=(0,0)))
model.add(MaxPooling2D(pool_size=(3,3), strides=(2,2)))

#L2
model.add(Convolution2D(nb_filter=128, nb_row=5, nb_col=5,
                        activation='relu',
                        subsample=(1,1)))
#model.add(ZeroPadding2D(padding=(0,0)))
model.add(MaxPooling2D(pool_size=(3,3), strides=(2,2)))

#L3
model.add(Convolution2D(nb_filter=256, nb_row=3, nb_col=3,
                        activation='relu',
                        subsample=(1,1)))
model.add(ZeroPadding2D(padding=(1,1)))
#L4
model.add(Convolution2D(nb_filter=256, nb_row=3, nb_col=3,
                        activation='relu',
                        subsample=(1,1)))
model.add(ZeroPadding2D(padding=(1,1)))
#L5
model.add(Convolution2D(nb_filter=256, nb_row=3, nb_col=3,
                        activation='relu',
                        subsample=(1,1)))
model.add(ZeroPadding2D(padding=(1,1)))
model.add(MaxPooling2D(pool_size=(3,3), strides=(2,2)))
#L6
model.add(Convolution2D(nb_filter=512, nb_row=7, nb_col=7,
                        activation='relu',
                        subsample=(1,1)))
#model.add(ZeroPadding2D(padding=(0,0)))

model.add(Dropout(0.5))
#L7
model.add(Convolution2D(nb_filter=512, nb_row=1, nb_col=1,
                        activation='relu',
                        subsample=(1,1)))
#model.add(ZeroPadding2D(padding=(0,0)))

model.add(Dropout(0.5))

model.add(Convolution2D(nb_filter=250, nb_row=1, nb_col=1,
                        activation='relu',
                        subsample=(1,1)))
#model.add(ZeroPadding2D(padding=(0,0)))
model.add(Flatten())
model.add(Activation("softmax"))
sgd = SGD(lr=0.003, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy',
              optimizer=sgd,
              metrics=['accuracy'])
checkpointer = ModelCheckpoint(filepath="checkpoint.hdf5", verbose=1, save_best_only=True)
tensorboard = TensorBoard(log_dir='logs', histogram_freq=0, write_graph=True, write_images=False)
model.fit_generator(datagen.flow(X_train, Y_train, batch_size=135), samples_per_epoch=18000, nb_epoch=nb_epoch,
          verbose=1, validation_data=(X_test, Y_test), callbacks=[checkpointer, tensorboard])`

**Comment by turambar** [comment]
@jadenyjw you can import an equivalent model by removing all of the zero padding layers and setting the `border_mode='same'` for convolutional layers whenever the zero padding is greater than 0 (layers 3-5).

Explanation: the zero padding in that model is strictly to ensure that the outputs for layers 3-5 have the same shape as the inputs (15 x 15). This is exactly what you get when you set `border_mode='same'`, which is supported by Keras (see [here](https://keras.io/layers/convolutional/)) with the TensorFlow backend (see [here](https://www.tensorflow.org/api_docs/python/nn/convolution)) and DL4J (see [here](https://github.com/deeplearning4j/deeplearning4j/blob/e31dcefe5a56c01a58cd9a8de51ead032ed610db/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ConvolutionMode.java)).

**Comment by turambar** [comment]
@tomthetrainer We should add the following to our model import documentation (i.e., ZeroPadding layers are not currently supported but folks should consider whether they can use `border_mode='same'` instead).

**Comment by jadenyjw** [comment]
Thank you so much @turambar, It seems to be accepting the model now. Another question I have is whether it will be able to preprocess (Featurewise Center and Featurewise stdNorm)?

**Comment by turambar** [comment]
@jadenyjw You mean preprocess the inputs? I don't see why not: we have a wide range of input preprocessing routines in DataVec ([here](https://github.com/deeplearning4j/DataVec/tree/master/datavec-api/src/main/java/org/datavec/api/transform/transform/doubletransform) are a few).

**Comment by rohrl** [comment]
@turambar Hi, I'm trying to import ResNet50 from Keras, but it contains ZeroPadding2D layer. I won't to import weights too, so I can't alter the model in a way that the weights will be incompatible (like remove the ZeroPadding2D layer). But I was thinking maybe I could replace it with a layer that just passes on the data without any change? What Keras layer could I use ? (it has to be importable by dl4j of course)

**Comment by turambar** [comment]
@rohrl just remove the zero padding layers and modify the Conv layers as appropriate. That won't interfere with weight import: the zero padding layers have no weights so they won't be listed in the weights file.

Your suggestion for a "do nothing" layer is not currently supported by DL4J or Keras AFAIK.

**Comment by rohrl** [comment]
@turambar Thanks, that worked.
However I got following exception further on:
```
Caused by: java.lang.ExceptionInInitializerError
	at org.deeplearning4j.nn.conf.NeuralNetConfiguration$Builder.build(NeuralNetConfiguration.java:1121)
	at org.deeplearning4j.nn.conf.ComputationGraphConfiguration$GraphBuilder.addLayer(ComputationGraphConfiguration.java:533)
	at org.deeplearning4j.nn.conf.ComputationGraphConfiguration$GraphBuilder.addLayer(ComputationGraphConfiguration.java:518)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraphConfiguration(KerasModel.java:395)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:450)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:439)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasModelAndWeights(KerasModelImport.java:205)
	at com.nuix.load.image.ModelImport.main(ModelImport.java:39)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
	... 5 more
Caused by: java.lang.TypeNotPresentException: Type org.nd4j.linalg.lossfunctions.impl.LossBinaryXENT not present
	at sun.reflect.annotation.TypeNotPresentExceptionProxy.generateException(TypeNotPresentExceptionProxy.java:46)
	at sun.reflect.annotation.AnnotationInvocationHandler.invoke(AnnotationInvocationHandler.java:84)
	at com.sun.proxy.$Proxy10.value(Unknown Source)
	at org.nd4j.shade.jackson.databind.introspect.JacksonAnnotationIntrospector.findSubtypes(JacksonAnnotationIntrospector.java:269)
	at org.nd4j.shade.jackson.databind.jsontype.impl.StdSubtypeResolver._collectAndResolve(StdSubtypeResolver.java:148)
	at org.nd4j.shade.jackson.databind.jsontype.impl.StdSubtypeResolver.collectAndResolveSubtypes(StdSubtypeResolver.java:111)
	at org.deeplearning4j.nn.conf.NeuralNetConfiguration.registerSubtypes(NeuralNetConfiguration.java:445)
	at org.deeplearning4j.nn.conf.NeuralNetConfiguration.configureMapper(NeuralNetConfiguration.java:373)
	at org.deeplearning4j.nn.conf.NeuralNetConfiguration.initMapper(NeuralNetConfiguration.java:363)
	at org.deeplearning4j.nn.conf.NeuralNetConfiguration.<clinit>(NeuralNetConfiguration.java:122)
	... 18 more
Caused by: java.lang.ClassNotFoundException: org.nd4j.linalg.lossfunctions.impl.LossBinaryXENT
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at sun.reflect.generics.factory.CoreReflectionFactory.makeNamedType(CoreReflectionFactory.java:114)
	at sun.reflect.generics.visitor.Reifier.visitClassTypeSignature(Reifier.java:125)
	at sun.reflect.generics.tree.ClassTypeSignature.accept(ClassTypeSignature.java:49)
	at sun.reflect.annotation.AnnotationParser.parseSig(AnnotationParser.java:439)
	at sun.reflect.annotation.AnnotationParser.parseClassValue(AnnotationParser.java:420)
	at sun.reflect.annotation.AnnotationParser.parseMemberValue(AnnotationParser.java:349)
	at sun.reflect.annotation.AnnotationParser.parseAnnotation2(AnnotationParser.java:286)
	at sun.reflect.annotation.AnnotationParser.parseAnnotation(AnnotationParser.java:226)
	at sun.reflect.annotation.AnnotationParser.parseAnnotationArray(AnnotationParser.java:765)
	at sun.reflect.annotation.AnnotationParser.parseArray(AnnotationParser.java:537)
	at sun.reflect.annotation.AnnotationParser.parseMemberValue(AnnotationParser.java:355)
	at sun.reflect.annotation.AnnotationParser.parseAnnotation2(AnnotationParser.java:286)
	at sun.reflect.annotation.AnnotationParser.parseAnnotations2(AnnotationParser.java:120)
	at sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:72)
	at java.lang.Class.createAnnotationData(Class.java:3521)
	at java.lang.Class.annotationData(Class.java:3510)
	at java.lang.Class.getDeclaredAnnotations(Class.java:3477)
	at org.nd4j.shade.jackson.databind.introspect.AnnotatedClass.resolveClassAnnotations(AnnotatedClass.java:308)
	at org.nd4j.shade.jackson.databind.introspect.AnnotatedClass.getAnnotation(AnnotatedClass.java:173)
	at org.nd4j.shade.jackson.databind.AnnotationIntrospector._findAnnotation(AnnotationIntrospector.java:1079)
	at org.nd4j.shade.jackson.databind.introspect.JacksonAnnotationIntrospector.findTypeName(JacksonAnnotationIntrospector.java:277)
	at org.nd4j.shade.jackson.databind.jsontype.impl.StdSubtypeResolver._collectAndResolve(StdSubtypeResolver.java:129)
	... 23 more
```

**Comment by agibsonccc** [comment]
That looks unrelated. /cc @AlexDBlack @eraly

**Comment by AlexDBlack** [comment]
Yeah, that looks like an issue with the ND4J install - or maybe mixed versions (using nd4j 0.7.2 with dl4j master, for example). I agree that it looks unrelated to keras import functionality or anything.

**Comment by rohrl** [comment]
Thanks, I rerun all builds and got rid of this exception.
But now it throws:
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.CommandLineWrapper.main(CommandLineWrapper.java:130)
Caused by: java.lang.IllegalStateException: Mis matched lengths: [32768] != [16384]
	at org.nd4j.linalg.util.LinAlgExceptions.assertSameLength(LinAlgExceptions.java:40)
	at org.nd4j.linalg.api.ops.BaseTransformOp.<init>(BaseTransformOp.java:48)
	at org.nd4j.linalg.api.ops.impl.transforms.Set.<init>(Set.java:25)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.assign(BaseNDArray.java:1203)
	at org.deeplearning4j.nn.layers.BaseLayer.setParam(BaseLayer.java:281)
	at org.deeplearning4j.nn.modelimport.keras.KerasLayer.copyWeightsToLayer(KerasLayer.java:432)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.helperCopyWeightsToModel(KerasModel.java:598)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:453)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:439)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasModelAndWeights(KerasModelImport.java:205)
	at com.nuix.load.image.ModelImport.main(ModelImport.java:43)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
	... 5 more
```

**Comment by turambar** [comment]
ZeroPadding2D layers are now explicitly supported in DL4J and in Keras model import. Included in master and in upcoming release.
